web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx

```diff
@@ -321,6 +325,44 @@ const Resources = memo(
       return cleanup;
     }, [onResourcesRefreshRequest, fetch]);
 
+    const { getAccessRequestButton } = props;
+
+    /**
+     * Keeps the list of `SharedUnifiedResource`, but also allows to return
+     * the original `UnifiedResourceResponse` object by the
+     * `SharedUnifiedResource['resource']`;
+     */
+    const mappedResources = useMemo(() => {
```

Comment:
Within this `useMemo` call, there's a lot of new terms being added (`mappedResources`, `originalResource`). For the `requestAccess` bulk action, we the argument to `action` `sharedResources`, but they're actually `SelectedResources`.

Here's my attempt at simplifying this a little bit, let me know what you think. I dropped any prefix other than `shared` and `selected`, so we have only three kinds now (resource, shared resource, selected resource). I also removed types where they did not have to be explicitly written out (because the types themselves are kinda complex already).

<details>
<summary>Patch</summary>

Copy then `pbpaste | git apply`.

```diff
diff --git a/web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx b/web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx
index 102804769e..a4a9574794 100644
--- a/web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx
+++ b/web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx
@@ -326,42 +326,42 @@ const Resources = memo(
     }, [onResourcesRefreshRequest, fetch]);
 
     const { getAccessRequestButton } = props;
+    // The action callback in the requestAccess action has access to
+    // `SharedUnifiedResource['resource']`, but `props.bulkAddResources` accepts
+    // `UnifiedResourceResponse`. Because of that, we need to to have the
+    // getUnifiedResourceFromSharedResource function.
+    const { sharedResources, getUnifiedResourceFromSharedResource } =
+      useMemo(() => {
+        const sharedResources: SharedUnifiedResource[] = [];
+        const sharedResourceToUnifiedResource = new Map<
+          SharedUnifiedResource['resource'],
+          UnifiedResourceResponse
+        >();
+
+        resources.forEach(resource => {
+          let sharedResource = mapToSharedResource(resource);
+          const accessRequestButton = getAccessRequestButton(resource);
+          if (accessRequestButton) {
+            sharedResource.ui.ActionButton = accessRequestButton;
+          }
 
-    /**
-     * Keeps the list of `SharedUnifiedResource`, but also allows to return
-     * the original `UnifiedResourceResponse` object by the
-     * `SharedUnifiedResource['resource']`;
-     */
-    const mappedResources = useMemo(() => {
-      // It is unfortunate that in the bulk action callback we get only
-      // `SharedUnifiedResource['resource']`, not `SharedUnifiedResource`.
-      // Because of that, we need to store SharedUnifiedResource list separately.
-      const list: SharedUnifiedResource[] = [];
-      const reverseMap = new Map<
-        SharedUnifiedResource['resource'],
-        UnifiedResourceResponse
-      >();
-
-      resources.forEach(originalResource => {
-        const { resource, ui } = mapToSharedResource(originalResource);
-
-        const sharedUnifiedResource: SharedUnifiedResource = {
-          resource,
-          ui: {
-            ActionButton:
-              getAccessRequestButton(originalResource) || ui.ActionButton,
-          },
+          sharedResources.push(sharedResource);
+          sharedResourceToUnifiedResource.set(
+            sharedResource.resource,
+            resource
+          );
+        });
+
+        const getUnifiedResourceFromSharedResource =
+          sharedResourceToUnifiedResource.get.bind(
+            sharedResourceToUnifiedResource
+          );
+
+        return {
+          sharedResources,
+          getUnifiedResourceFromSharedResource,
         };
-        list.push(sharedUnifiedResource);
-        reverseMap.set(sharedUnifiedResource.resource, originalResource);
-      });
-      return {
-        list,
-        getOriginalResource: (
-          r: SharedUnifiedResource['resource']
-        ): UnifiedResourceResponse => reverseMap.get(r),
-      };
-    }, [resources, getAccessRequestButton]);
+      }, [resources, getAccessRequestButton]);
 
     const resourceIds =
       props.userPreferences.clusterPreferences?.pinnedResources?.resourceIds;
@@ -390,10 +390,10 @@ const Resources = memo(
                   Icon: icons.AddCircle,
                   text: 'Request Access',
                   disabled: false,
-                  action: sharedResources =>
+                  action: selectedResources =>
                     props.bulkAddResources(
-                      sharedResources.map(sharedResource =>
-                        mappedResources.getOriginalResource(
+                      selectedResources.map(sharedResource =>
+                        getUnifiedResourceFromSharedResource(
                           sharedResource.resource
                         )
                       )
@@ -414,7 +414,7 @@ const Resources = memo(
             ? props.integratedAccessRequests.availabilityFilter
             : undefined
         }
-        resources={mappedResources.list}
+        resources={sharedResources}
         resourcesFetchAttempt={attempt}
         fetchResources={fetch}
         availableKinds={[

```
</details>